apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java

```diff
@@ -0,0 +1,392 @@
+        Optional<List<McsTag>> tags = Optional.of(checkAndProcessMcsClaim(mcsClaim));
+        persistTags(tags, entityManager);
+      }
+      default -> throw new RuntimeException("Unknown claim type.");
+    }
+  }
+
+  /**
+   * Persists the tags to the database.
+   *
+   * @param tags List of tags to persist.
+   * @param entityManager the EntityManager.
+   * @param <TTag> Generic type of the tags.
+   */
+  private <TTag> void persistTags(Optional<List<TTag>> tags, EntityManager entityManager) {
+    if (tags.isPresent()) {
+      for (TTag tag : tags.get()) {
+        entityManager.merge(tag);
+      }
+    }
+  }
+
+  /**
+   * Checks for SAMHSA codes in an MCS claim and constructs the tags.
+   *
+   * @param mcsClaim The claim to check.
+   * @return A list of tag entities to persist.
+   */
+  private List<McsTag> checkAndProcessMcsClaim(RdaMcsClaim mcsClaim) {
+    Optional<List<TagDetails>> entries = getPossibleMcsSamhsaFields(mcsClaim);
+    if (entries.isPresent()) {
+      List<McsTag> mcsTags = new ArrayList<>();
+      mcsTags.add(
+          McsTag.builder()
+              .claim(mcsClaim.getIdrClmHdIcn())
+              .code(TagCode._42CFRPart2.toString())
+              .details(entries.get())
+              .build());
+      mcsTags.add(
+          McsTag.builder()
+              .claim(mcsClaim.getIdrClmHdIcn())
+              .code(TagCode.R.toString())
+              .details(entries.get())
+              .build());
+      return mcsTags;
+    }
+    return Collections.emptyList();
+  }
+
+  /**
+   * Checks for SAMHSA codes in a FISS Claim and constructs the tags.
+   *
+   * @param fissClaim The claim to check.
+   * @return A list of tag entities to persist.
+   */
+  private List<FissTag> checkAndProcessFissClaim(RdaFissClaim fissClaim) {
+    Optional<List<TagDetails>> entries = getPossibleFissSamhsaFields(fissClaim);
+    if (entries.isPresent()) {
+      List<FissTag> fissTags = new ArrayList<>();
+      fissTags.add(
+          FissTag.builder()
+              .claim(fissClaim.getClaimId())
+              .code(TagCode._42CFRPart2.toString())
+              .details(entries.get())
+              .build());
+      fissTags.add(
+          FissTag.builder()
+              .claim(fissClaim.getClaimId())
+              .code(TagCode.R.toString())
+              .details(entries.get())
+              .build());
+      return fissTags;
+    }
+    return Collections.emptyList();
+  }
+
+  /**
+   * Constructs a list of TagDetail objects for an MCS claim.
+   *
+   * @param mcsClaim The claim to check.
+   * @return a list of TagDetail objects, one for each SAMHSA code found in the claim.
+   */
+  private Optional<List<TagDetails>> getPossibleMcsSamhsaFields(RdaMcsClaim mcsClaim) {
+    List<TagDetails> entries = new ArrayList<>();
+    LocalDate serviceDate =
+        mcsClaim.getIdrHdrFromDateOfSvc() == null
+            ? LocalDate.parse("1970-01-01")
+            : mcsClaim.getIdrHdrFromDateOfSvc();
```

Comment:
Might be worth checking with @alex-dzeda to make sure we're using the right dates. I recall there was some nuance to this when we were discussing this previously.

Reply:
The safest option is to imagine a time period (eg start and end date). If any of those dates is after an applicable start date, I'd consider it to be within range.